### ISSUE TITLE
ISO between Modifiables and Immutables

### ISSUE THREAD

**Issue by SimY4** [immutables/immutables#1250]
I have an idea floating around about writing a code that will provide you immutable handles for something that internally requires modifiable type. One concrete example of that would be DynamoDB mapper which only can work with bean friendly modifiables.

The proposal I wanted to make is to add an extra module, that will have a single interface definition like:

```
interface Iso<Immutable, Modifiable> {
  Immutable fromModifiable(Modifiable modifiable);

  Modifiable fromImmutable(Immutable immutable);
}
```

If present on classpath, and your value definition contain both immutable and modifiable annotations, it'll generate an instance of Iso for them:

```
@Value.Immutable
@Value.Modifiable
interface MyEntity {
  String getString();
}

//-- generates ImmutableMyEntity
//-- generates ModifiableMyEntity
//-- generates Iso:

final class MyEntityIso<ImmutableMyEntity, ModifiableMyEntity> {
  @Override
  public ImmutableMyEntity fromModifiable(ModifiableMyEntity modifiable) {
    return modifiable.toImmutable();
  }

  @Override
  public ModifiableMyEntity fromImmutable(ImmutableMyEntity immutable) {
    return ModifiableMyEntity.from(immutable);
  }
}
```

Is it something you are interested in?

**Comment by elucash** [comment]
I wonder how it will be used. Will it be auto discovered? so far what I see is probably useful, but brings custom API. If you ever will instantiate or reference an instance of `MyEntityIso<ImmutableMyEntity, ModifiableMyEntity>` it seems that it's not a much of a leap from `Iso.of(ModifiableMyEntity::toImmutable, ModifiableMyEntity::from)` you can figure implementation, I guess. Probably it would make sense to generate singleton static instance of such Iso/lens on `ModifiableMyEntity.immutableIso()`, but then we might need an API module as you mentioned or shoehorn into smth like `Map.Entry<Function<M, I>, Function<I, M>>` which is a bit ugly but no additional API needed.

**Comment by asereda-gs** [comment]
Yes probably something like [Converter](https://guava.dev/releases/23.0/api/docs/com/google/common/base/Converter.html) from guava.

Will reflection be too slow for this ?

**Comment by SimY4** [comment]
@SimY4 the cool thing about `Modifiable.toImmutable()` and `Modifiable.from(immutable)` is that the handle deep nesting. And they have the right type signatures for reflection based frameworks. In example:

```
Immutable { 
  field: Immutable
  listField: List<Immutable>
}
```
is converted to
```
Modifiable { 
  field: Modifiable 
  listField: List<Modifiable>
}
```

Which I find awesome. :) I'm currently working on a proof of concept that will allow me to use this to make DynamoDBMapper work with deep nested bean friendly copy of Immutable type. Most of the things I reported before was from that proof of concept. I can share my findings here when it's ready.

**Comment by asereda-gs** [comment]
But you can still create `Converter<ImmutableMyEntity, ModifiableMyEntity>` (see [Converter API](https://guava.dev/releases/23.0/api/docs/com/google/common/base/Converter.html)) using a factory:

```
static <A, B> Converter<A, B> create(A.class, B.class) {
   // use reflection to call from(...) / toImmutable()
}
```
or directly from two functions (as Eugene mentioned previously) :  [Converter.from(Function, Function)](https://guava.dev/releases/23.0/api/docs/com/google/common/base/Converter.html#from-com.google.common.base.Function-com.google.common.base.Function-)

This way `ISO` (or `Converter`) doesn't need to be code-generated.

Usually we try to avoid generating too many classes.

**Comment by SimY4** [comment]
Alright, it took me a while but I have something to share now. I created a small repo where I explore the possibility of using Immutables with DynanoDBMapper (repo is located in here: https://github.com/SimY4/ImmutablesDynamoDBMapperPOC)

Points of interest are these three definitions:
[Entity](https://github.com/SimY4/ImmutablesDynamoDBMapperPOC/blob/main/src/main/java/com/github/simy4/poc/model/Entity.java) <- top level DynamoDB document
[Address](https://github.com/SimY4/ImmutablesDynamoDBMapperPOC/blob/main/src/main/java/com/github/simy4/poc/model/Address.java) <- nested document
[Email](https://github.com/SimY4/ImmutablesDynamoDBMapperPOC/blob/main/src/main/java/com/github/simy4/poc/model/Email.java) <- another nested document

The repetitive part in all of them is this converter which I'd like to abstract over. But unfortunately, I can't.
Restrictions being that I need that converter to obey certain rules. Rules that I can't break without using some nasty reflection to discover Immutable and Modifiable companions.

But if these transitions (Immutable <-> Modifiable) will be abstracted in an interface. This whole duplication will go away.

**Comment by elucash** [comment]
Is the issue is that DynamoDB SDK annotations requires a class? (so `Converter.from(Function, Function)` cannot be easily used)?
One step closer can be something like the following

```java
final class Converter extends AbstractConverter<ModifiableEmail, ImmutableEmail> {
   public Converter() {
     super(e -> new ModifiableEmail.from(e), ModifiableEmail::toImmutable());
   }
}
```

Definitely we could generate something like this, the problem is that what you need is that generated code would use the converter API from DynamoDB SDK. And while we have precedents with some stuff generated for the libs (like Jackson or MongoDB client), the more niche we go, the more we introduce code/features that very small fraction of users will ever use.

Given that you control the style and can rely on the naming conventions of your choice, is it possible to just write separate annotation processor which can, for example, hook onto `@Data` annotation and generate trivial converter class. It could be plain annotation processor and, let's say, _JavaPoet_ java source builder or just string builder (to not bother with templating engines).
